src/common/mfu_daos.c

```diff
@@ -1313,3 +1326,2840 @@ int mfu_flist_copy_daos(
+    if (path_len >= FILENAME_LEN) {
+        MFU_LOG(MFU_LOG_ERR, "filename is too long");
+        rc = 1;
+        goto out;
+    }
+
+    printf("Serializing Container to %s\n", filename);
+    /* keep track of number of files written */
+    (*files_written)++;
+
+    /* init HDF5 datatypes in HDF5 file */
+    rc = init_hdf5_file(&hdf5, filename);
+    if (rc != 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to init hdf5 file");
+        rc = 1;
+        goto out;
+    }
+    hdf5.oid_data = MFU_MALLOC(num_oids * sizeof(oid_t));
+    if (hdf5.oid_data == NULL) {
+        rc = ENOMEM;
+        goto out;
+    }
+    hdf5.dkey_data = MFU_MALLOC(sizeof(dkey_t));
+    if (hdf5.dkey_data == NULL) {
+        rc = ENOMEM;
+        mfu_free(&hdf5.oid_data);
+        goto out;
+    }
+    hdf5.akey_data = MFU_MALLOC(sizeof(akey_t));
+    if (hdf5.akey_data == NULL) {
+        rc = ENOMEM;
+        mfu_free(&hdf5.oid_data);
+        mfu_free(&hdf5.dkey_data);
+        goto out;
+    }
+    hdf5.dk = &(hdf5.dkey_data);
+    hdf5.ak = &(hdf5.akey_data);
+    hdf5.oid = &(hdf5.oid_data);
+
+    /* size is total oids for this rank, loop over each oid and serialize */
+    for (i = 0; i < num_oids; i++) {
+        /* open DAOS object based on oid to get obj
+         * handle
+         */
+        daos_obj_id_t oid;
+        oid.hi = mfu_flist_file_get_oid_high(flist, i);
+        oid.lo = mfu_flist_file_get_oid_low(flist, i);
+        (*hdf5.oid)[i].oid_hi = oid.hi;
+        (*hdf5.oid)[i].oid_low = oid.lo;
+        rc = daos_obj_open(da->src_coh, oid, 0, &oh, NULL);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "failed to open object: %d", rc);
+            goto out;
+        }
+        rc = serialize_dkeys(&hdf5, &dk_index, &ak_index,
+                             &total_dkeys, &total_akeys,
+                             &oh, &i);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "failed to serialize keys: %d", rc);
+            goto out;
+        }
+        /* close source and destination object */
+        rc = daos_obj_close(oh, NULL);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "failed to close object: %d", rc);
+            goto out;
+        }
+    }
+
+    /* write container version as attribute */
+    rc = cont_serialize_version(&hdf5, version);
+    if (rc != 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to serialize version");
+        goto out;
+    }
+
+    rc = cont_serialize_usr_attrs(&hdf5, da->src_coh);
+    if (rc != 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to serialize user attributes");
+        goto out;
+    }
+
+    rc = cont_serialize_props(&hdf5, da->src_coh);
+    if (rc != 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to serialize cont layout");
+        goto out;
+    }
+
+    /* sum files_written across all ranks to get total */
+    uint64_t total_files_written = 0;
+    MPI_Allreduce(files_written, &total_files_written, 1, MPI_UNSIGNED,
+               MPI_SUM, MPI_COMM_WORLD);
```

Comment:
@dsikich All/most of the goto statements above this reduction might potentially cause this to hang if only some of the ranks error.